File: factorial_calculator.rb
# frozen_string_literal: true

# Модуль для вычисления факториала
module FactorialCalculator
  # Вычисляет факториал числа n (n!).
  # Возвращает 1 для n <= 1.
  # Для n > 1 рекурсивно вычисляет n * (n-1)!
  #
  # @param n [Integer] Неотрицательное целое число
  # @return [Integer] Факториал числа n
  def self.factorial(n)
    return 1 if n <= 1
    n * factorial(n - 1)
  end
end

# Пример использования
puts FactorialCalculator.factorial(5) # => 120

File: factorial_calculator.py
import math

def factorial(n):
    """
    Calculate the factorial of a non-negative integer n.

    Args:
        n (int): A non-negative integer.

    Returns:
        int: The factorial of n.

    Raises:
        ValueError: If n is negative.
    """
    if n < 0:
        raise ValueError("Factorial is not defined for negative numbers.")
    # Using math.prod for Python 3.8+ which is efficient for calculating products
    # Fallback to a loop for older versions or for clarity using math.factorial
    return math.factorial(n)

# Example usage:
if __name__ == "__main__":
    try:
        number = 5
        result = factorial(number)
        print(f"The factorial of {number} is {result}.")
    except ValueError as e:
        print(e)
